File: fieldmark_py/fieldmark/tests/normalize_html.py
"""Shared HTML normalization helper for cross-stack snapshot tests."""

from __future__ import annotations

import re

_COMMENT_RE = re.compile(r"<!--.*?-->", re.DOTALL)
_WHITESPACE_RE = re.compile(r"\s+")
# strips <input type="hidden" ...> so CSRF tokens don't break cross-stack parity diffs
_HIDDEN_INPUT_RE = re.compile(r'<input[^>]+type="hidden"[^>]*>', re.IGNORECASE)


def normalise_component(html: str) -> str:
    """Strip HTML comments, collapse all whitespace to single spaces, trim."""
    html = _COMMENT_RE.sub("", html)
    html = html.replace("&#34;", "&quot;")
    html = _WHITESPACE_RE.sub(" ", html).strip()
    return html


def normalise_for_parity(html: str) -> str:
    """Like normalise_component but also strips hidden inputs so CSRF tokens
    do not break cross-stack snapshot comparisons."""
    html = _HIDDEN_INPUT_RE.sub("", html)
    return normalise_component(html)


def extract_variant(example_content: str, variant_name: str) -> str:
    """Extract a named variant block from a canonical component example file.

    Blocks are delimited by ``<!-- variant: <name> ... -->`` comment lines.
    Returns the normalised content between that delimiter and the next one.
    """
    lines = example_content.splitlines()
    start_marker = f"<!-- variant: {variant_name}"
    in_block = False
    collected: list[str] = []

    for line in lines:
        stripped = line.rstrip()
        if stripped == f"{start_marker} -->" or stripped.startswith(f"{start_marker} "):
            in_block = True
            continue
        if in_block and stripped.startswith("<!-- variant:"):
            break
        if in_block:
            collected.append(stripped)

    return normalise_component("\n".join(collected))

File: fieldmark-go/internal/web/testutil/normalizehtml.go
// Package testutil provides shared helpers for web-layer snapshot tests.
package testutil

import (
	"regexp"
	"strings"
)

var (
	commentRe    = regexp.MustCompile(`(?s)<!--.*?-->`)
	whitespaceRe = regexp.MustCompile(`\s+`)
	// hiddenInputRe strips <input type="hidden" ...> elements added by CSRF frameworks
	// so parity snapshot comparisons are not broken by per-request tokens.
	hiddenInputRe = regexp.MustCompile(`(?i)<input[^>]+type="hidden"[^>]*>`)
)

// NormaliseComponent strips HTML comments, collapses whitespace runs to a
// single space, and trims the result. Used for cross-stack snapshot comparison.
func NormaliseComponent(html string) string {
	html = commentRe.ReplaceAllString(html, "")
	html = strings.ReplaceAll(html, "&#34;", "&quot;")
	html = whitespaceRe.ReplaceAllString(html, " ")
	return strings.TrimSpace(html)
}

// NormaliseForParity calls NormaliseComponent and additionally strips
// <input type="hidden"> elements so CSRF tokens do not break cross-stack diffs.
func NormaliseForParity(html string) string {
	html = hiddenInputRe.ReplaceAllString(html, "")
	return NormaliseComponent(html)
}

// ExtractVariant extracts the content of a named variant block from a
// canonical component example file and returns it normalised.
// Blocks are delimited by <!-- variant: <name> ... --> comment lines.
func ExtractVariant(exampleContent, variantName string) string {
	startMarker := "<!-- variant: " + variantName
	lines := strings.Split(exampleContent, "\n")
	inBlock := false
	var collected []string

	for _, line := range lines {
		trimmed := strings.TrimRight(line, " \t\r")
		if trimmed == startMarker+" -->" || strings.HasPrefix(trimmed, startMarker+" ") {
			inBlock = true
			continue
		}
		if inBlock && strings.HasPrefix(trimmed, "<!-- variant:") {
			break
		}
		if inBlock {
			collected = append(collected, trimmed)
		}
	}

	return NormaliseComponent(strings.Join(collected, "\n"))
}
